### DISCUSSION TITLE
Inconsistent behavior of the Required validation rules (incl. unless, if, and other related Required rules)

### DISCUSSION THREAD

**Discussion by davidkvasnovsky** [spatie/laravel-data#511] [General]
Let's assume we have a Data object, where the `properties` field (for example, a JSON column casted as Data) is both nullable and required only when the User role differs from Customer.

```php
class CreateUserData extends Data
{
    public function __construct(
        public string $name,
        public string $role,
        #[RequiredUnless(field: 'role', values: ['customer'])]
        public ?DataObject $properties,
    ) {
    }
}
```

Now when we try to create `CreateUserData` from the request with the following payload (without the `properties` field):

```json
{
    "role": "editor",
    "name": "Some Name"
}
```
CreateUserController:

```php
class CreateUserController
{
    public function __invoke(CreateUserData $data): CreateUserData
    {
        return $data;
    }
}

```

Everything works, and a new instance is created with the following response:

```json
{
    "name": "Some Name",
    "role": "editor",
    "properties": null
}
```

But I think that the expected behavior should be that it won't pass validation. If we add validation rules manually:

```php
class CreateUserData extends Data
{
    public function __construct(
        public string $name,
        public string $role,
//        #[RequiredUnless(field: 'role', values: ['customer'])]
        public ?DataObject $properties,
    ) {
    }

    public static function rules(ValidationContext $context): array
    {
        return [
            'properties' => 'required_unless:role,customer',
        ];
    }
}
```

It behaves as expected and throws a validation error.

```json
{
    "message": "The properties field is required unless role is in customer.",
    "errors": {
        "properties": [
            "The properties field is required unless role is in customer."
        ]
    }
}
```

Same applies for all related Required validation rules (`Required`, `Required Unless`, `Required If`...).

Edited: I just recently found out that the same bug applies for `#[Rule]` attribute.

**Comment by rubenvanassche** [comment]
Yeah we have some problems on that front with the attributes, using custom rules is the solution for now or see: https://github.com/spatie/laravel-data/issues/500. Someone is already trying to fix it for optional over there. It should probably also happen for nullable data objects.
